From 54843af899318b26ea19d11572354339825d5bd7 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Date: Fri, 7 Dec 2012 12:06:47 +0100
Subject: [PORT FROM MAIN] softap: Prevent TOCTOU and symlink attack

BZ: 73435

A TOCTOU attack was found in the SoftapController and a patch was provided
upstream by Google. This patch is a port of the fix on SoftapControllerTI.cpp
which diverged from SoftapController.cpp.

Feature   : Wlan
Category  : Improvement
Purpose   : AOSP+
Origin    : Upstream
Upstream-candidate: no

Change-Id: If93433a179f4d18bc0cabb15f10e0e00ab99290c
Orig-Change-Id: I17ee624aa2a4557e8aa87f58d5429dac45febb3d
Signed-off-by: Quentin Casasnovas <quentinx.casasnovas@intel.com>
Reviewed-on: http://android.intel.com:8080/83931
Reviewed-by: Naoumenko, PaulX <paulx.naoumenko@intel.com>
Reviewed-by: Khenissi, MaherX <maherx.khenissi@intel.com>
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Tested-by: Zurmely, PierreX <pierrex.zurmely@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 SoftapControllerTI.cpp |   18 ++++++++++++++----
 1 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/SoftapControllerTI.cpp b/SoftapControllerTI.cpp
index ea20003..b81ec76 100644
--- a/SoftapControllerTI.cpp
+++ b/SoftapControllerTI.cpp
@@ -448,9 +448,16 @@ int SoftapController::setSoftap(int argc, char *argv[]) {
        return -1;
     }
 
-    FILE* fp2 = fopen(HOSTAPD_CONF_FILE, "w");
+    int fd = open(HOSTAPD_CONF_FILE, O_CREAT | O_TRUNC | O_WRONLY | O_NOFOLLOW, 0660);
+    if (fd < 0) {
+       LOGE("Softap set - hostapd.conf file read failed");
+       fclose(fp);
+       return -1;
+    }
+    FILE* fp2 = fdopen(fd, "w");
     if (!fp2) {
        LOGE("Softap set - hostapd.conf file read failed");
+       close(fd);
        fclose(fp);
        return -1;
     }
@@ -549,23 +556,26 @@ int SoftapController::setSoftap(int argc, char *argv[]) {
     fputs(bufptr, fp2);
 
     fclose(fp);
-    fclose(fp2);
 
 	/* Note: apparently open can fail to set permissions correctly at times */
-    if (chmod(HOSTAPD_CONF_FILE, 0660) < 0) {
+    if (fchmod(fd, 0660) < 0) {
         ALOGE("Error changing permissions of %s to 0660: %s",
                 HOSTAPD_CONF_FILE, strerror(errno));
+        fclose(fp2);
         unlink(HOSTAPD_CONF_FILE);
         return -1;
     }
 
-    if (chown(HOSTAPD_CONF_FILE, AID_SYSTEM, AID_WIFI) < 0) {
+    if (fchown(fd, AID_SYSTEM, AID_WIFI) < 0) {
         ALOGE("Error changing group ownership of %s to %d: %s",
                 HOSTAPD_CONF_FILE, AID_WIFI, strerror(errno));
+        fclose(fp2);
         unlink(HOSTAPD_CONF_FILE);
         return -1;
     }
 
+    fclose(fp2);
+
     memcpy(mBackupBuf,buf,sizeof(buf));
 
     // we take the wakelock here because the stop/start is lengthy
-- 
1.7.4.1

